pennylane/circuit_graph.py

```diff
@@ -206,6 +211,124 @@ def hash(self):
+                operations, also include the gates that diagonalize the
+                measured wires such that they are in the eigenbasis of the circuit observables.
+
+        Returns:
+            str: OpenQASM serialization of the circuit
+        """
+        # We import decompose_queue here to avoid a circular import
+        from pennylane.qnodes.base import decompose_queue  # pylint: disable=import-outside-toplevel
+
+        # add the QASM headers
+        qasm_str = "OPENQASM 2.0;\n"
+        qasm_str += 'include "qelib1.inc";\n'
+
+        if self.num_wires == 0:
+            # empty circuit
+            return qasm_str
+
+        # create the quantum and classical registers
+        qasm_str += "qreg q[{}];\n".format(self.num_wires)
+        qasm_str += "creg c[{}];\n".format(self.num_wires)
+
+        # Map PennyLane gate names to equivalent QASM gate names.
+        # Note that QASM has two native gates:
+        #
+        # - U (equivalent to qml.U3)
+        # - CX (equivalent to qml.CNOT)
+        #
+        # All other gates are defined in the file qelib1.inc:
+        # https://github.com/Qiskit/openqasm/blob/master/examples/generic/qelib1.inc
+        #
+        # Question: do other implementations of QASM exist that
+        # don't use the qelib1.inc gate definitions?
+        # Perhaps we allow developers to pass their own native_qasm_gates dictionary
+        # to this method, which is used instead of the default below
+        # if provided.
+        native_qasm_gates = {
+            "CNOT": "cx",
+            "CZ": "cz",
+            "U3": "u3",
+            "U2": "u2",
+            "U1": "u1",
+            "Identity": "id",
+            "PauliX": "x",
+            "PauliY": "y",
+            "PauliZ": "z",
+            "Hadamard": "h",
+            "S": "s",
+            "S.inv": "sdg",
+            "T": "t",
+            "T.inv": "tdg",
+            "RX": "rx",
+            "RY": "ry",
+            "RZ": "rz",
+            "CRX": "crx",
+            "CRY": "cry",
+            "CRZ": "crz",
+            "SWAP": "swap",
+            "Toffoli": "ccx",
+            "CSWAP": "cswap",
+            "PhaseShift": "u1",
+        }
+
+        # get the user applied circuit operations
+        operations = self.operations
+
+        if rotations:
+            # if requested, append diagonalizing gates corresponding
+            # to circuit observables
+            operations += self.diagonalizing_gates
+
+        class QASMDevice:
+            """A mock device, to be used when performing the decomposition.
+            The short_name is used in error messages if the decomposition fails.
+            """
+
+            # pylint: disable=too-few-public-methods
+            short_name = "QASM serializer"
+            supports_operation = staticmethod(lambda x: x in native_qasm_gates)
+
+        # decompose the queue
+        operations = decompose_queue(operations, QASMDevice)
```

Comment:
Nitpick: `decomposed_ops` for naming the left-hand side?

Reply:
```suggestion
        decomposed_ops = decompose_queue(operations, QASMDevice)
```